bigtop-packages/src/charm/zookeeper/layer-zookeeper/reactive/zookeeper.py

```diff
@@ -52,36 +66,176 @@ def update_network_interface():
+#
+# 0. Whenever the Zookeeper server starts, it attempts to determine
+#    whether it is the Zookeeper leader. If so, it sets a flag on the
+#    Juju peer relation data.
+#
+# 1. When a node is added or remove from the cluster, the Juju leader
+#    runs `check_cluster`, and generates a queue of nodes in the
+#    cluster, with the Zookeeper lead node sorted last in the
+#    queue. It also sets a nonce, to identify this restart queue
+#    uniquely, and thus handle the situation where another node is
+#    added or restarted while we're still reacting to the first node's
+#    addition or removal. The leader drops the queue and nonce into
+#    the leadership data as "restart_queue" and "restart_nonce",
+#    respectively.
+#
+# 2. When any node detects a leadership.changed.restart_queue event,
+#    it runs `restart_for_quorum`, which is a noop unless the node's
+#    private address is the first element of the restart queue. In
+#    that case, if the node is the Juju leader, it will restart, then
+#    remove itself from the restart queue, triggering another
+#    leadership.changed.restart_queue event. If the node isn't the
+#    Juju leader, it will restart itself, then run `inform_restart`,
+#    from the zookeeper-quorum interface. That triggers a little
+#    dance:
+#
+# 2a. `inform_restart` will set
+#     "{relation-name}.restarted.<nonce>". This will trigger a
+#     relation data changed event.
+#
+# 2b. When the relation data changes, `toggle_restarted` in the
+#     zookeeper-quorum interface will trigger.
+#
+# 2c. `toggle_restarted` is a noop on all nodes save the Juju
+#     leader. If a node is the Juju leader, and one of its relations
+#     has set zkpeer.restarted.<nonce> for the current nonce, the
+#     leader will set the zkpeer.restarted state on itself.
+#
+# 3. If a leader is in the 'zkpeer.restarted' state, it will run
+#    `update_restart_queue`. This will remove the first node in the
+#    restart_queue, clear the 'zkpeer.restarted' state, then update
+#    the restart_queue leadership data, looping us back to step 2.
+#
+# 4. Once all the nodes have restarted, we should be in the following state:
+#
+#    * All nodes have an updated Zookeeper server running with the new
+#    * peer data.
+#
+#    * The Zookeeper leader has restarted last, which should help
+#      prevent orphaned jobs, per the Zookeeper docs.
+#
+#    * The Juju leader is no longer in the zkpeer.restarted state
+#
+#    * peers still have zkpeer.restarted.<nonce> set on their relation
+#      data. This is okay, as we will generate a new nonce next time,
+#      and the data is small.
+#
+# Edge cases and potential bugs:
+#
+# 1. Juju leader changes in the middle of a restart: this needs to be
+#    tested, but should work. The leader data should get picked up by
+#    the next leader, and we'll need to generate a fresh restart_queue
+#    anyway, as the Juju leader node has gone away.
```

Comment:
Darn it. I tested this, and it doesn't work. If you remove the leader using remove-unit, it still attempts to wrangle everything as leader while it is tearing itself down. It fails to get RelationBase, throws an exception, and gets stuck in a state where it can't finish tearing down.

The good news is that I can fix that bug by just telling the leader to only run check_cluster when zkpeer.joined is set. But that causes other bugs (the new leader never sets up the restart queue). 

Continuing to troubleshoot. Will push a fix once I have it ...